Sources/NIOCore/EventLoopFuture+WithEventLoop.swift

```diff
@@ -41,25 +41,12 @@ extension EventLoopFuture {
     /// - returns: A future that will receive the eventual value.
     @inlinable
     @preconcurrency
-    public func flatMapWithEventLoop<NewValue>(_ callback: @escaping @Sendable (Value, EventLoop) -> EventLoopFuture<NewValue>) -> EventLoopFuture<NewValue> {
-        let next = EventLoopPromise<NewValue>.makeUnleakablePromise(eventLoop: self.eventLoop)
-        self._whenComplete { [eventLoop = self.eventLoop] in
-            switch self._value! {
-            case .success(let t):
-                let futureU = callback(t, eventLoop)
-                if futureU.eventLoop.inEventLoop {
-                    return futureU._addCallback {
-                        next._setValue(value: futureU._value!)
-                    }
-                } else {
-                    futureU.cascade(to: next)
-                    return CallbackList()
-                }
-            case .failure(let error):
-                return next._setValue(value: .failure(error))
-            }
-        }
-        return next.futureResult
+    public func flatMapWithEventLoop<NewValue: Sendable>(_ callback: @escaping @Sendable (Value, EventLoop) -> EventLoopFuture<NewValue>) -> EventLoopFuture<NewValue> {
+        // Is this the same thing and still fast?
+        let eventLoop = self.eventLoop
+        return self.flatMap {
+            callback($0, eventLoop)
+        }.hop(to: self.eventLoop)
```

Comment:
you can delete the `.hop(to:)`. `self.flatMap` will return a future on `self.eventLoop`

Reply:
and the previous implementation probably also allocates less, depends on the optimiser a bit. Why not just revert?

Reply:
Right. Well this method became a lot easier then :D